ws.c

```diff
@@ -221,34 +226,54 @@ get_ws_frame(unsigned char *in_buffer, int buf_len, unsigned char **payload_ptr,
 	if (length_field <= 125) {
 		payload_len = length_field;
 	} else if (length_field == 126) { /* msglen is 16bit */
+		uint16_t tmp16;
 		if (buf_len < 4)
 			return INCOMPLETE_DATA;
-		payload_len = ntohs(*(uint16_t *)(in_buffer + 2));
+		memcpy(&tmp16, in_buffer + pos, 2);
+		payload_len = ntohs(tmp16);
 		pos += 2;
 	} else if (length_field == 127) { /* msglen is 64bit */
+		int i;
+		uint64_t tmp64 = 0;
 		if (buf_len < 10)
 			return INCOMPLETE_DATA;
-		payload_len = ntohs(*(uint64_t *)(in_buffer + 2));
-		pos += 8;
+		/* swab bytes from big endian to host byte order */
+		for (i = 56; i >= 0; i -= 8) {
+			tmp64 |= (uint64_t)in_buffer[pos++] << i;
+		}
```

Comment:
`bswap_64` can be used instead. It looks pretty standard, thought not sure about win32, but if it is not supported there it will be better to introduce a helper.

And we should not do the byte swap if we are a big endian platform

Reply:
- This code is a no-op on big-endian, since it is platform neutral to swap bytes in this way. On big endian the result will just be the original ordering.  Operator << and >> on C/C++ are defined in such a way such that they are arithmetic operators (increasing/decreasing by a power of 2). So, on big endian this is a no-op.
- `bswap_64` doesn't appear to exist on my macOS system.  I can do a follow-up PR with adding bswap_64 etc if you like ...

Reply:
Oh, sorry, I've just read the comment, not the code, and I missed that you are simply reading from the `char *` as network byte order to host byte order. So let's update the comment then.

But the code is wrong, `(uint64_t)in_buffer[pos++]` will read `uint64` from the `in_buffer` always, while you need to read only one byte - `(uint64_t)(data[pos++] & 0xff)`

Also let's use `CHAR_BIT` instead of `8`, and instead of `56` something like `8 * (CHAR_BIT - 1)`. Or, actually I guess we cannot use `sizeof(uint64_t)`, since it should be `8` on any platform?

And let's introduce helper for this and write a unit test.

Reply:
> sizeof(uint64_t), since it should be 8 on any platform?

Well .. yes and no.  Yes I believe these are specified to be *exactly* 64-bits.. now.. if `CHAR_BIT` differs.. you are in for a surprising time!  Since `sizeof(uint64_t)` might not be 8.. it will instead be something like `(64/CHAR_BIT)` 

Hmm.. yes I see what you are saying about the constants.  

Remember this though: POSIX mandates `CHAR_BIT == 8.` I know you compile for technically-non-POSIX systems (win32) -- but even there CHAR_BIT is 8. I think you should make it a project assumption that CHAR_BIT == 8 and require that.  I suspect more than just `ws.c` would explode in that case if CHAR_BIT != 8.  


---

So: Really.. What happens on a hypothetical machine that has CHAT_BIT not set to 8?  Remember on such a machine also `sizeof(whatever)` is affected because `sizeof` is defined as returning the number of characters. 

- If `CHAR_BIT == 6`, for instance, then we would need to adjust this loop.. as you suggested... to be based off of `CHAR_BIT` for the initial conditions and loop decrement.. and things should work.
- If `CHAR_BIT == 12` we can't even do `& 0xff` as you suggested since that would be wrong! We would need to be doing `& 0x7ff` here!  
- Actually in all cases you want to *not* be doing ANY `& 0xff` <-- this has, baked-in, an assumption about `CHAR_BIT == 8` :)

You see what I'm saying?  This project already is assuming `CHAR_BIT == 8` implicitly or even explicitly everywhere. 

I must say this again -- you ought to make it a project assumption that `CHAR_BIT == 8`. `ws.c` already would fail on such a system if you scroll up and down in the file it already assumes CHAR_BIT == 8 in code I didn't write in the file.

So long story short: Yes, I can adjust it as you say -- but I really do think this project should assume `CHAR_BIT == 8` for its own sanity..

Reply:
... Also remember the protocol is specifying that exactly 8 * 8-bit-bytes of data are coming in off the wire here.. literal constants like 8 and 64 are baked-in to the protocol already.. so there's that too.

Reply:
Now.. I could be persuaded to use `CHAR_BIT` here for *readability*... to make people reading this code more easily understand *intent* here.

.. but this project really does not support `CHAR_BIT != 8`, I can assure you of that..

Reply:
Yes, I see your points, thanks. So let's just add a check into cmake & autotools that `CHAR_BIT` is 8, and done with this.